site/topics/testing/testing.rst

```diff
@@ -165,12 +165,57 @@ Square of Sums Example Tests
 Celsius to Fahrenheit Example Tests
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+ .. code-block:: python
+    :linenos:
+
+    def celsius_to_fahrenheit(temp_in_celsius: float) -> float:
+        """
+        Convert a temperature from Celsius units to Fahrenheit units.
+
+        :rtype: float
+        :param temp_in_celsius: The temperature in Celsius to be converted.
+        :return: The temperature in Fahrenheit.
+        """
+        partial_conversion = temp_in_celsius * 9 / 5
+        temp_in_fahrenheit = partial_conversion + 32
+        return temp_in_fahrenheit
+
+
+    # Tests for celsius_to_fahrenheit function
+    assert 32 == celsius_to_fahrenheit(0)
+    assert -40 == celsius_to_fahrenheit(-40)
+    assert 86 == celsius_to_fahrenheit(30)
+    assert 89.6 == celsius_to_fahrenheit(32)
+    # To address precision issues, we can look for a sufficiently small difference between the expected and actual
+    assert 0.001 > abs(celsius_to_fahrenheit(37.7777) - 100)
+
+
+* Above is a series of assertion tests for the ``celsius_to_fahrenheit`` function
+* Notice the key, broad tests
+
+    * Input of zero
+    * Negative input
+    * Positive input
+    * Input is a float
+    * Output is a float
```

Comment:
I'd argue `-40` is a special case and does not cover the general negative number case. It's the only point where the two scales are equal, which sounds like an edge case to me.

_Test driven development_ says you actually write some failing tests first, before writing the code that makes the tests pass. Once the tests all pass, either you add more failing tests and iterate to get the full behaviour you want, or you're done. I'm not one to say anyone should follow TDD (I don't enjoy inflicting pain on others), but there is a benefit to at least thinking about tests _first_. Thinking about the problem you're solving before there's a solution, I'll just assert, is very helpful.

Not really applicable here, but I also say that tests are the first clients of your code. If you find something hard to test, it's likely hard to use.

Reply:
> Not really applicable here, but I also say that tests are the first clients of your code. If you find something hard to test, it's likely hard to use.

![image](https://user-images.githubusercontent.com/1328048/162767931-48a411da-ccd4-4a1f-b897-7eeb80d36f95.png)